lib/arpv4.ml

```diff
@@ -175,18 +197,30 @@ module Make (Ethif : V1_LWT.ETHIF) = struct
     if Hashtbl.mem t.cache ip then (
       Hashtbl.find t.cache ip
     ) else (
-      let response, waker = MProf.Trace.named_wait "ARP response" in
-      (* printf "ARP query: %s -> [probe]\n%!" (Ipaddr.V4.to_string ip); *)
-      Hashtbl.add t.cache ip response;
-      Hashtbl.add t.pending ip waker;
-      (* First request, so send a query packet *)
-      output_probe t ip >>= fun () ->
-      response
+      let rec try_query t ip counter : (Macaddr.t option) Lwt.t = 
+        Hashtbl.remove t.pending ip; (* responses are too little, too late *)
+        Hashtbl.remove t.cache ip; (* new attempts to resolve should not be referred
+                                        to the previous sleeping thread *)
+        match counter with 
+        | 0 -> return None
+        | n ->
+          let response, waker = MProf.Trace.named_wait "ARP response" in
+          Hashtbl.add t.cache ip response;
+          Hashtbl.add t.pending ip waker; 
+          output_probe t ip >>= fun () -> 
+          Time.sleep probe_repeat_delay >>= fun () ->
```

Comment:
If I'm reading this right, we always wait for `probe_repeat_delay` after the ARP request is sent, even if we get a reply sooner. I think this means that the first user to request an address will always have to wait (other callers may get the answer from the cache even while the first caller is still waiting).

Also, any caller waiting on the initial cached `response` will keep waiting if it fails, even if there is a successful retry.

I'd suggest putting the retry logic on the inside of the loop, so the entry in `t.cache` represents the result of the whole lookup-with-retry operation, not just the result of the current try. Then you don't have to modify the cache on each retry.

I haven't tested it, though, so could be talking nonsense...